File: control/reactor.lua
local state = require("control/state")
local settings = require("control/settings")

---@param reactor LuaEntity
local function engageReactor(reactor)
    reactor.set_recipe('warp-reactor-1')
end

---@param reactor LuaEntity
local function setStartingRecipe(reactor)
    if settings.getWarpzoneGracePeriodTicks(state.currentWarpzone()) ~= 0 then
        reactor.set_recipe(nil)
        return
    end
    engageReactor(reactor)
end

local function onInit(surface)
    local reactor = surface.create_entity({
        name = "warp-reactor",
        position = { x = 2, y = -2 },
        direction = defines.direction.north,
        force = game.forces.player
    })
    state.setCurrentReactor(reactor)
    setStartingRecipe(reactor)
end

---@param event EventData.on_entity_cloned
local function onCloned(event)
    state.setCurrentReactor(event.destination)
    setStartingRecipe(event.destination)
end

---@param event EventData.on_object_destroyed
local function onDestroyed(event)
    if not state.isReactorDestroyed(event) then return end
    game.set_lose_ending_info({ title = {"warp-ending.reactor-destroyed"}, message={"warp-ending.reactor-destroyed-message"} })
    game.set_game_state({ can_continue = false, game_finished = true, player_won = false })
end

return { onInit = onInit, onCloned = onCloned, onDestroyed = onDestroyed, engageReactor = engageReactor }

File: control/reactor.ts
import { entities, warpReactorRecipes } from "constants"
import { LuaEntity, LuaSurface, OnEntityClonedEvent, OnObjectDestroyedEvent } from "factorio:runtime"
import { getWarpzoneGracePeriodTicks } from "control/settings";
import { currentWarpzone, isCurrentReactorDestroyed, setCurrentReactor } from "control/state";

export function engageReactor(reactor: LuaEntity) {
    reactor.set_recipe(warpReactorRecipes[0].name);
}

export function setStartingRecipe(reactor: LuaEntity) {
    if (getWarpzoneGracePeriodTicks(currentWarpzone()) != 0) {
        reactor.set_recipe();
        return;
    }

    engageReactor(reactor);
}

export function onInit(surface: LuaSurface) {
    const reactor = surface.create_entity({
        name: entities.warpReactor,
        position: { x: 2, y: -2 },
        direction: defines.direction.north,
        force: game.forces.player
    })!;

    setCurrentReactor(reactor);
    setStartingRecipe(reactor);
}

export function onCloned(event: OnEntityClonedEvent) {
    setCurrentReactor(event.destination);
    setStartingRecipe(event.destination);
}

export function onDestroyed(event: OnObjectDestroyedEvent) {
    if (!isCurrentReactorDestroyed(event)) return;

    game.set_lose_ending_info({ title: ["warp-ending.reactor-destroyed"], message: ["warp-ending.reactor-destroyed-message"] });
    game.set_game_state({ can_continue: false, game_finished: true, player_won: false })
}
